### PR TITLE
ALTAPPS-802: Shared review serializable models

### PR DESCRIPTION
**Pull request by vladkash**
**YouTrack Issues**:
[#ALTAPPS-802](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-802)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Reviewed and fixed models fields types
- Removed unnecessary fields

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import kotlinx.serialization.Serializable
 @Serializable
 data class CommentStatisticsEntry(
     @SerialName("thread")
-    val thread: CommentThread,
+    val thread: CommentThread? = null,
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt:9]: Backend can send an unknown value in the `thread` field. We should handle it properly with default value. This is why `thread` should be nullable. The alternative for it is a default enum value like `UNKNOWN`.

File: shared/src/commonMain/kotlin/org/hyperskill/app/stages/domain/model/Stage.kt [hunk 2]
```diff
@@ -22,7 +18,7 @@ data class Stage(
     val projectStagesCount: Int,
 
     @SerialName("step")
-    val stepId: Long,
+    val stepId: Long?,
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/stages/domain/model/Stage.kt:21]: Could you please add a link to the backend code, to verify types nullability.

**Review by XanderZhu — CHANGES_REQUESTED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/stages/domain/model/Stage.kt (hunk 2 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/stages/domain/model/Stage.kt:21]: Screenshot 2023-06-05 at 1 41 54 PM

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt (hunk 1 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt:9]: > Backend can send an unknown value in the thread field. We should handle it properly with default value. This is why thread should be nullable.

Yes

> The alternative for it is a default enum value like UNKNOWN.

I think a null is better in case of backend returns an unknown thread type for the mobile client. Nullable forces us to unwrap it before using it, but with a non-nullable value we can forget to check for the `UNKNOWN` type.

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/view/mapper/CommentThreadTitleMapper.kt
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step/view/mapper/CommentThreadTitleMapper.kt:1]: It's better to not change this file. The calling side should check for `commentThread == null`

File: shared/src/commonMain/kotlin/org/hyperskill/app/likes/domain/model/LikeValue.kt [hunk 3]
```diff
@@ -1,6 +1,5 @@
 package org.hyperskill.app.likes.domain.model
 
 enum class LikeValue(val value: Int) {
-    VALUE_0(0),
     VALUE_1(1)
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/likes/domain/model/LikeValue.kt:4]: `VALUE_1 ` is used when we call `abuseComment`. Let's add a human-readable comment about what it means, for example "Means dislike".

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/ViewData/StepViewDataMapper.swift [hunk 4]
```diff
@@ -48,6 +48,6 @@ final class StepViewDataMapper {
             count: commentStatisticsEntry.totalCount
         )
 
-        return StepCommentStatisticViewData(id: commentStatisticsEntry.thread.name, title: title)
+        return StepCommentStatisticViewData(id: commentStatisticsEntry.thread?.name ?? title, title: title)
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/ViewData/StepViewDataMapper.swift:51]: If `commentStatisticsEntry.thread == nil` then we don't want to display this in the UI. It's better to add an early `guard` statement and return a `nil` value.

^^ If this will introduce a lot of changes in the current code, please let me know and we will think what to do.

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt (hunk 1 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt:9]: In my opinion, the `UNKNOWN` enum value is better because it presents human readable value and clearly shows the state of the `CommentThread`.  We just need to check all the values in the `when` expression.

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/profile/view/fragment/ProfileFragment.kt [hunk 5]
```diff
@@ -342,12 +342,7 @@ class ProfileFragment :
         with(viewBinding.profileFooterLayout) {
             if (profile.country != null) {
                 profileAboutLivesTextView.text =
-                    "${resources.getString(org.hyperskill.app.R.string.profile_lives_in_text)} ${
-                    Locale(
-                        Locale.ENGLISH.language,
-                        profile.country!!
-                    ).displayCountry
-                    }"
+                    "${resources.getString(org.hyperskill.app.R.string.profile_lives_in_text)} ${profile.country!!}"
```
[the remark this answers is not in the document]
Comment (vladkash) [reply androidHyperskillApp/src/main/java/org/hyperskill/app/android/profile/view/fragment/ProfileFragment.kt:345]: Yes, I remembered this task then looked for usages and decided to fix it here

File: shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt (hunk 1 above)
Comment (vladkash) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step/domain/model/CommentStatisticsEntry.kt:9]: @XanderZhu @ivan-magda Let's discuss that on next PBR, now we have different approaches in the project. We will choose the one and create task to technical debt

File: androidHyperskillApp/src/main/java/org/hyperskill/app/android/profile/view/fragment/ProfileFragment.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivan-magda) [reply androidHyperskillApp/src/main/java/org/hyperskill/app/android/profile/view/fragment/ProfileFragment.kt:345]: Nice 👍

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/ViewData/StepViewData.swift [hunk 6]
```diff
@@ -10,7 +10,7 @@ struct StepViewData {
 
     var text: String
 
-    let commentsStatistics: [StepCommentStatisticViewData]
+    let commentsStatistics: [StepCommentStatisticViewData?]
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/Step/ViewData/StepViewData.swift:13]: Instead, it's better to use `compactMap` over `map` when mapping data:
```swift
func mapStepToViewData(_ step: Step) -> StepViewData {
    ...

    step.commentsStatistics.compactMap(self.mapCommentStatisticsEntryToViewData(_:))

    ...
}
```

File: shared/src/commonMain/kotlin/org/hyperskill/app/likes/domain/model/LikeValue.kt [hunk 7]
```diff
@@ -1,6 +1,5 @@
 package org.hyperskill.app.likes.domain.model
 
 enum class LikeValue(val value: Int) {
-    VALUE_0(0),
-    VALUE_1(1)
+    ABUSE_COMMENT(1)
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/likes/domain/model/LikeValue.kt:4]: Technically `VALUE_1` can be used with different subjects and depends on the subject that value means different things, so `ABUSE_COMMENT` is only true when a subject is COMMENT.

Screenshot 2023-06-06 at 1 55 33 PM

I propose you revert the change and earlier I asked you to just add some documentation comment about when `VALUE_1` can be used in the project.
